### DISCUSSION TITLE
Using tl from C++

### DISCUSSION THREAD

**Discussion by yuriv** [teal-language/tl#477] [Q&A]
Could I use tl from C++? I am using sol2 C++ library to work with lua from C++. Could I "upgrade" lua with tl?

**Comment by lenscas** [comment]
I never used C++, let alone C++ together with lua/teal but as far as I know you can ask lua to run the tl.tl script and then load in a teal file through require.

I do the same thing for my Rust project and it works fine. There I basically run
```lua
local tl =(function()
//entirety of the tl.tl script
end)()
tl.loader()
require "path/to/tl/file/I/want/to/load"
```

**Reply by yuriv** [reply]
Yes, but I am speaking about lua function which returns typed values to C++/Rust (doesn't matter). For instance:

`local tl =(function() : number
return 3.14
end)()`

Is tl var will be of number type?

**Reply by lenscas** [reply]
`number` is just a normal lua type. So, if that function gets translated to lua (for example because you loaded in the teal compiler) then every way to interact with lua should be able to deal with it.

teal only runs at compile time, the signature of a function gets removed after it runs. So `local tl =(function() : number return 3.14 end)()` simply runs as `local tl =(function()  return 3.14 end)()`

**Reply by yuriv** [reply]
Ok, I understood. This is a specific problem of the sol2 library. It retrieves values from lua like sum type (because pure lua has no strong typing) and I need to cast to the explicit type manually on the client side. If I want to work with tl & sol2 I will need to add some support for Teal in sol2.

If it's not so hard, post your tl initialization code snippet (in rust).

Thanks

**Reply by lenscas** [reply]
it is a macro that does it as it is part of my rust <-> teal FFI library, but literally all it does is first create a string like
```rust
let primed_vm_string = format!(
        "local tl = (function()\n{}\nend)()\ntl.loader()\n",
        compiler
    );
    ```
    and then does a 
    ```rust
                format!("{}\n return require('{}')",#primed_vm_string,require)
    ```
    to get the final result that will be passed to lua.

https://github.com/lenscas/tealr/blob/5a96267f4536a4173b9267add108959b3e70714d/tealr_derive/src/lib.rs#L287
